Read_group: A New Hope

ryv-odoo opened the pull request:
## Rationale

The method `read_group()` was designed to be used by the web client to efficiently compute aggregations grouped by one or more fields. However, more and more developers have been using it from the backend to make computations more efficient (avoid doing the aggregation in Python). Unfortunately, the API was designed for the web client, which adds a lot of boilerplate when used in the Python (list of dict with misleading key names).

Method `_read_group()` was created to improve the performance of `read_group()` for backend use (4ef0c00b4b37d75b78c5cd30dba004ffd0f8b0ac), but didn't change the API and based the implementation on `read_group()` itself.

This rewrites `_read_group()` from scratch with a new API to make it easier to use from the backend. Also, split the method to make it easy to override and add custom behavior.

## New API of `_read_group()`

The purpose of the method is unchanged: get field aggregations specified by `aggregates` grouped by the given `groupby` fields where records are filtered by the `domain`.

### Parameters

- Rename `fields` into `aggregates`; parameter `aggregates` is a list of _explicit_ aggregate specifications (no implicit aggregator coming from `group_operator`). 
- Parameter `groupby` becomes the second parameter (just before `aggregates`). It is much more readable for the returning value. Also the groupby specification for date/datetime fields should be always explicit (no default 'month').
- New parameter `having`, which is a domain-like list to filter the result depending of aggregate values. It allows to simplify some code and avoid to fetch unnecessary data from the database.
- Parameter `orderby` is renamed into `order`, to be consistent with method `search()`. Order terms should be an explicit aggregate specification or groupby specification (+ ASC/DESC).
- Remove parameter `lazy`, which was only used for the webclient and doesn't have any sense to be used in backend.

### Returned value

The returned value is a list of tuples containing, in order, the group values and aggregate values (flatten). For instance, if you group by `['foo', 'bar']` and aggregate `['baz:sum', 'quux:max']`, the result will contain tuples like `(foo_value, bar_value, baz_sum_value, quux_max_value)`.

If a group is relational field, its corresponding value will be a recordset (with a correct prefetch set that includes all returned values). For other fields, the returned values are similar to the ones you get from a record: no labels anymore, date/datetime values are not transformed strings, NULL values are returned as `False`, etc.

## Performance

Minor performance improvements:
- The aggregate `__count` generates `COUNT(*)` instead `COUNT(<table>.id)` (which generates useless null checks from PostgreSQL). Moreover, `COUNT(*)` is only added when we explicitly ask for `__count`.
- Remove the implicit `min(<table>.id)` aggregate added previously to the query at each read_group.
- The recordsets with correct prefetch set are returned for relational group values: it adds a very small cost of instantiation, but avoids manual browsing and bad field prefetching in business code.
- Don't include extra values generated for the webclient: `__domain`, `__range`, `__fold`, `__context`.
- A falsy domain simply doesn't make any query at all, just like `search`. It is very convenient with domains like `[(..., 'in', self.ids)]` where `self.ids` can be empty (it happens a lot in case of onchange methods).

Performance regression: 
- The values of aggregation `array_agg` are now ordered by `id`, which is less efficient but it ensures a deterministic result. 

### Performance Testing

Performance tests show that the new version/usage is faster in general.
With a populated database (`--size=medium` + some tweaks) and with all these changes,
the performance of the `_read_group`/`read_group` calls, for 2280 calls, is in average +- 9.17 % faster.

- 388 (on 2280) cases are statically slower (mean = -7.66 %, median = -5.17%) than before for several reasons:
      - more restricted access due to the commit named "simplify security of read_group", which in some case generates extra queries.
      - small overhead to instantiate recordsets;
      - `array_agg` is now deterministic;
      - noise.
- 862 cases are statically faster (mean = +181.08 %, median = +15.82%) than before.


## Next step

The second part of this task will be focused on the API of public method `read_group()`:
- Make it more consistent with this new `_read_group`.
- Simplify the parsing of data (always return raw values aside the labelled ones).
- Avoid useless aggregations.

Enterprise: https://github.com/odoo/enterprise/pull/38639

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/110737).

AntoineVDV commented:
@ryv-odoo Documentation PR? :pray:

ryv-odoo commented on odoo/models.py:
```diff
@@ -1731,19 +1768,326 @@ def clear_caches(cls):
+
+    @api.model
+    def _read_group_check_field_access_rights(self, field_names):
+        # TODO: should have a general hook (not only for read_group) for this kind of stuff
+        self.check_field_access_rights('read', field_names)
```
Yep, but I cannot restrict in `check_field_access_rights` (or check with call stack whatever I came from `_read_group`) because it is used by `_read` (and we don't want restrict for these cases).

ryv-odoo commented on odoo/models.py:
```diff
@@ -1731,19 +1768,326 @@ def clear_caches(cls):
+        # TODO: should have a general hook (not only for read_group) for this kind of stuff
+        self.check_field_access_rights('read', field_names)
+
+    @api.model
```
IMO, `@api.model` means `doesn't rely on self` which is the case here. (In the same way than _search has also `@api.model`)

ryv-odoo commented on odoo/models.py:
```diff
@@ -1731,19 +1768,326 @@ def clear_caches(cls):
+        self.check_field_access_rights('read', field_names)
+
+    @api.model
+    def _read_group(self, domain, groupby=(), aggregates=(), having=(), offset=0, limit=None, order=None):
```
IMO, because of `groupby` and `aggregates` has disjoint specification, both are distinguishable. Then I don't want to force keyword.

ryv-odoo replied:
> Also performing the validation upstream would obviate the need to test if fnames_to_flush before checking the fields ACLs.

I need it also for the count(*) sadly :/

ryv-odoo commented on odoo/models.py:
```diff
@@ -1731,19 +1768,326 @@ def clear_caches(cls):
+                browse_partial = functools.partial(Model, env=self.env, prefetch_ids=prefetch_ids)
+                return lambda value: browse_partial(ids=(value,)) if value else empty_value
+
+            return lambda value: value if value is not None else empty_value
```
Nice catch :+1:

ryv-odoo commented on odoo/models.py:
```diff
@@ -1731,19 +1768,326 @@ def clear_caches(cls):
+
+        self.env.cr.execute('\n' + '\n'.join(query_parts), query_params)
+
+        return list(zip(*(
+            self._read_group_postprocess(spec, values)
+            for spec, values in zip(itertools.chain(groupby, aggregates), zip(*self.env.cr.fetchall()))
+        )))
```
> The double transposition and 3-deep zip is pretty rough.

Yes, you are right. It is too complicate to read. I will try to find something else.

> So if I understand correctly this basically converts the output of GROUP_BY into a column-major form, post-processes each column according to its spec, then transposes back to row-major for the final output?

Not only the `GROUP_BY` but also the aggregates selected.

> Since _read_group_postprocess just maps over the values anyway, wouldn't it be easier (and more readable) to create a sequence of post-processing functions then apply those to the group rows? Unless there are plants to leverage the column-major traversal but that seems iffy.

Yep, I will try to have a more readable code.

ryv-odoo commented on odoo/models.py:
```diff
@@ -1886,8 +2230,8 @@ def _read_group_fill_temporal(self, data, groupby, aggregated_fields, annotated_
         :param list data: the data containing groups
-        :param list groupby: name of the first group by
-        :param list aggregated_fields: list of aggregated fields in the query
+        :param list groupby: list field group on
```
`???` for my English or the change ? Because it was a list before, the doc was wrong.

std-odoo commented on odoo/addons/test_read_group/tests/test_private_read_group.py:
```diff
@@ -0,0 +1,518 @@
+                    ["Donkey Kong"],
+                ),
+            ],
+        )
```
could be greate to add some test on SQL injection attempt
```
(like `_read_group([], [], [name:array_agg OR])`)
```

std-odoo commented on odoo/models.py:
```diff
@@ -74,11 +74,18 @@
+def split_read_group_spec(spec: str) -> 'tuple[str]':
+    res_match = regex_aggregate_spec.match(spec)
+    if not res_match:
+        raise ValueError(f'Invalid aggregate/groupby specification {spec!r}. Valid specification example: "field:agg"')
```
can't we translate the error message here ? 
(same below)

ryv-odoo replied:
IMO, we should never translate developer error. User should never see it (because he cannot construct it by itself).

std-odoo commented on odoo/models.py:
```diff
@@ -74,11 +74,18 @@
 regex_pg_name = re.compile(r'^[a-z_][a-z0-9_$]*$', re.I)
 regex_field_agg = re.compile(r'(\w+)(?::(\w+)(?:\((\w+)\))?)?')
 
+regex_aggregate_spec = re.compile(r'(\w+)(?::(\w+))?$')
```
IDK if it could lead to SQL injection, but just to be sure
```suggestion
regex_aggregate_spec = re.compile(r'^(\w+)(?::(\w+))?$')
```

ryv-odoo replied:
I use `re.match` then the `^` is already implicitly added.

std-odoo replied:
not a fan of that being implicit but ok

std-odoo commented on odoo/models.py:
```diff
@@ -1731,6 +1738,310 @@ def clear_caches(cls):
+
+        field_expression = self._inherits_join_calc(self._table, fname, query)
+        if field.type == 'datetime' and self.env.context.get('tz') in pytz.all_timezones_set:
+            # TODO: '{self._context.get('tz')}' should be passed as args
```
if we still use tz from context, can we add an assert with a regex ? (in case one day they update the lib and we have a timezone like `Los Angele's` with a single quote)

ryv-odoo replied:
I prefer not, I don't like regex. We will create a compositional SQL String, it will solve this issue cleanly.

std-odoo commented on odoo/models.py:
```diff
@@ -1731,19 +1768,325 @@ def clear_caches(cls):
+            return orderby_terms, extra_groupby_terms, fnames_used
+
+        for order_term in order.split(','):
+            order_field, order_orientation = order_term.split()
```
this line crash with
```py
self.env['res.partner']._read_group([], ['name'])
```

shouldn't we have one by default ?

ryv-odoo replied:
It was a small regression done this morning. Already fix.

std-odoo commented on odoo/models.py:
```diff
@@ -1731,19 +1768,326 @@ def clear_caches(cls):
+        if fname not in self:
+            raise ValueError(f"Invalid field {fname!r} on model {self._name!r} for {aggregate_spec!r}.")
+        if func not in VALID_AGGREGATE_FUNCTIONS:
+            raise ValueError(f"Invalid aggregate method {func!r} for {aggregate_spec!r}.")
```
can we have a more explicit message if no func are given ? (so we know it's not a bug, but it should be explicit)

> Aggregate method is mandatory

std-odoo commented on odoo/models.py:
```diff
@@ -1731,19 +1768,326 @@ def clear_caches(cls):
+
+                op = 'AND' if leaf == '&' else 'OR'
+                stack_op.append(f'({condition_1} {op} {condition_2})')
+            elif isinstance(leaf, (list, tuple)):
```
could be nice to add a test about my `'"="'` trick if you remember

ryv-odoo replied:
Already done in the `test_malformed_params`

std-odoo reviewed (commented):
@ryv-odoo Hi :) 

Some small comments / propositions

Happy to see this cleaner read_group :)

ryv-odoo commented on addons/mrp/report/stock_forecasted.py:
```diff
@@ -44,16 +44,16 @@ def _get_report_header(self, product_template_ids, product_ids, wh_location_ids)
-        grouped_moves = self.env['stock.move'].read_group(move_domain, ['product_qty:sum'], 'product_id')
-        res['draft_production_qty']['out'] = sum(move['product_qty'] for move in grouped_moves)
+        [product_qty] = self.env['stock.move']._read_group(move_domain, aggregates=['product_qty:sum'])[0]
+        res['draft_production_qty']['out'] = product_qty or 0.0
```
Yes :'(

rco-odoo commented on odoo/models.py:
```diff
@@ -1731,6 +1738,310 @@ def clear_caches(cls):
+        orderby_terms, extra_groupby_terms, fnames_used = self._read_group_orderby(query, groupby, order, order_traverse_many2one)
+        fnames_to_flush.update(fnames_used)
+
+        from_clause, where_clause, query_params = query.get_sql()
```
you may use `query.select(*(group_by_terms + select_terms))`

rco-odoo commented on addons/crm_iap_mine/models/crm_iap_lead_mining_request.py:
```diff
@@ -98,15 +98,12 @@ def _compute_tooltip(self):
+            [('lead_mining_request_id', 'in', self.ids)],
+            ['lead_mining_request_id'], ['__count'])
+        mapped_data = {
+            lead_mining_request: count
```
should be `lead_mining_request.id`

rco-odoo commented on addons/fleet/models/fleet_vehicle_model.py:
```diff
@@ -54,9 +54,9 @@ def name_get(self):
     def _compute_vehicle_count(self):
         group = self.env['fleet.vehicle']._read_group(
-            [('model_id', 'in', self.ids)], ['id', 'model_id'], groupby='model_id', lazy=False,
+            [('model_id', 'in', self.ids)], 'model_id', aggregates=['__count'],
```
suggestion: `['model_id']`

william-andre commented on addons/account/models/account_account.py:
```diff
@@ -26,13 +26,13 @@ def _check_reconcile(self):
-            fields=['company_id', 'ids:array_agg(id)'],
             groupby=['company_id'],
+            aggregates=['id:array_agg'],
+            having=[('__count', '>', 1)],
```
Why not
```suggestion
            having=[('count', '>', 1)],
```

william-andre replied:
What about `id:count` then?
I just find it weird the have a special case

william-andre reviewed (commented):
Very nice :+1: 

Just a little detail:

> The values of aggregation `array_agg` are now ordered by `id`, which is less efficient but it ensures a deterministic result.

Note that this is not accurate. It might be more deterministic, but it is not ensured: in my experience, `array_agg` is often used to aggregate ids. If we write those ids on a x2m field, we might not have them in the same order as what was written because the ORM will fetch the values using the `_order` of the comodel.
So in a flow that might sometimes but not always trigger a flush, the order of the `ids` might be the order of the `_read_group`, or the order of the model (because the ORM will keep the order of the values given to `write` or `__set__` in the cache until invalidated).

ryv-odoo commented:
Thank you for your input @william-andre 

However, I don't understand your point about the `array_agg`:

> It might be more deterministic, but it is not ensured

What do you mean ? `id` is always unique and not-null, then it is deterministic. Example of SQL expression generated: `ARRAY_AGG("test_read_group_task"."name" ORDER BY "test_read_group_task"."id")`. Or I miss something ?

> the order of the ids might be the order of the _read_group

It doesn't make sense: the order of the read_group orders groups, but the order of array_agg should order inside a group it-self.

> or the order of the model 

It will generate extra join in case of many2one (also some `_order` of models are even not deterministic).

rco-odoo commented on addons/analytic/models/analytic_account.py:
```diff
@@ -138,22 +138,19 @@ def _read_group(self, domain, groupby=(), aggregates=(), having=(), offset=0, li
 
-        new_aggregates = ['id:array_agg', *(agg for agg in aggregates if agg not in SPECIAL)]
-        new_result = super()._read_group(domain, groupby, new_aggregates, having, offset, limit, order)
+        new_aggregates = ['id:recordset', *(agg for agg in aggregates if agg not in SPECIAL)]
+        result = super()._read_group(domain, groupby, new_aggregates, having, offset, limit, order)
```
The former name was chosen on purpose: `new_result` corresponded to `new_aggregates`, while `result` corresponded to `aggregates`.

ryv-odoo replied:
Ok I see, I made this change because we are looping on `result` instead of `new_result` just setting `result = []`.
Then I am currently fixing the issue without change naming.

rco-odoo reviewed (approved):
Excellent!

rco-odoo commented:
@robodoo rebase-ff r+

robodoo commented:
Merge method set to rebase and fast-forward.

robodoo commented:
@ryv-odoo @rco-odoo linked pull request(s) odoo/documentation#4064, odoo/enterprise#38639 not ready. Linked PRs are not staged until all of them are ready.

AntoineVDV commented on odoo/models.py:
```diff
@@ -1741,19 +1790,362 @@ def clear_caches(cls):
+
+        :param list domain: :ref:`A search domain <reference/orm/domains>`. Use an empty
+                list to match all records.
+        :param list groupby: list of groupby descriptions by which the records will be grouped.
+                A groupby description is either a field (then it will be grouped by that field)
+                or a string 'field:granularity'. Right now, the only supported granularities
+                are 'day', 'week', 'month', 'quarter' or 'year', and they only make sense for
+                date/datetime fields.
+        :param list aggregates: list of aggregates specification.
+                Each element is 'field:agg' (aggregate field with aggregation function 'agg').
+                The possible aggregation functions are the ones provided by
+                `PostgreSQL <https://www.postgresql.org/docs/current/static/functions-aggregate.html>`_,
+                'count_distinct' with the expected meaning and 'recordset' to act like 'array_agg'
+                converted into a recordset.
```
IMO it'd make sense to use the literal markup on string arguments so that it's clear that you're documenting the expected values.

![image](https://user-images.githubusercontent.com/15896813/233054641-399d1511-abe2-41de-92b8-ec75223978c0.png)

```suggestion
        :param list groupby: list of groupby descriptions by which the records will be grouped.
                A groupby description is either a field (then it will be grouped by that field)
                or a string `'field:granularity'`. Right now, the only supported granularities
                are `'day'`, `'week'`, `'month'`, `'quarter'` or `'year'`, and they only make sense for
                date/datetime fields.
        :param list aggregates: list of aggregates specification.
                Each element is `'field:agg'` (aggregate field with aggregation function `'agg'`).
                The possible aggregation functions are the ones provided by
                `PostgreSQL <https://www.postgresql.org/docs/current/static/functions-aggregate.html>`_,
                `'count_distinct'` with the expected meaning and `'recordset'` to act like `'array_agg'`
                converted into a recordset.
```

AntoineVDV commented on odoo/models.py:
```diff
@@ -1741,19 +1790,362 @@ def clear_caches(cls):
+        :param int limit: optional max number of groups to return
+        :param str order: optional ``order by`` specification, for
+                overriding the natural sort ordering of the groups,
+                see also :py:meth:`~osv.osv.osv.search`.
```
I assume you meant to create a link to the method here?
```suggestion
                see also :meth:`~.search`.
```

jjscarafia commented:
Hi @ryv-odoo. Can you share any script fo apply this changes on custom code? (for eg. in OCA modules)
Thanks!

ryv-odoo commented:
> Hi @ryv-odoo. Can you share any script fo apply this changes on custom code? (for eg. in OCA modules) Thanks!

I made a little script to help me when I changed the codebase of Odoo: 
https://github.com/ryv-odoo/odoo_scripts/blob/master/read_group_replacement.py

But of course it doesn't work perfectly (and it's not very clean). Because I changed the returned values, you still have to check every use of it.

Note that the old `read_group` (public method) doesn't change its behavior much (but it's not very efficient to use it in the backend).
